File: Going Dutch/MCMailComposer.m
//
//  MCMailComposer.m
//  We all pay
//
//  Created by Mark Cornelisse on 14-05-14.
//  Copyright (c) 2014 Mark Cornelisse. All rights reserved.
//

#import "MCMailComposer.h"

#import "MCReturnPayment.h"
#import "MCPayment+addons.h"
#import "MCPerson+addons.h"

@implementation MCMailComposer

#pragma mark - Private methods.



#pragma mark - New in this class.

- (NSArray *)getMailAddresses
{
    NSArray *sortDescriptorArray = @[[NSSortDescriptor sortDescriptorWithKey:@"dateCreated" ascending:YES]];
    NSArray *allPeople = [[_tonightsBill peoplePresent] sortedArrayUsingDescriptors:sortDescriptorArray];
    // Create a list of all email addresses
    NSMutableArray *listOfMailAddresses = [[NSMutableArray alloc] init];
    for (MCPerson *person in allPeople) {
        if ([person defaultEmailAddress]) {
            [listOfMailAddresses addObject:[person defaultEmailAddress]];
        }
    }
    return listOfMailAddresses;
}

- (NSString *)getSubject
{
    NSString *subject1 = NSLocalizedString(@"EMAIL_SUBJECT_PART_ONE", @"Bill overview of our trip to %@");
    return [[NSString alloc] initWithFormat:@"%@ %@.", subject1, [[self tonightsBill] tripName]];
}

- (NSString *)getMailBody
{
    // Generate the text for the email.
    NSMutableString *mailBody = [[NSMutableString alloc] init];
    NSNumberFormatter *nf = [[NSNumberFormatter alloc] init];
    [nf setNumberStyle:NSNumberFormatterCurrencyStyle];
    [mailBody appendString:@"https://itunes.apple.com/us/app/we-all-pay/id642135963?mt=8&uo=4\n\n"];
    
    [mailBody appendFormat:@"%@ %@,\n", NSLocalizedString(@"EMAIL_DEAR", @"Just Dear as in \"Dear Mark\""), [[self tonightsBill] stringOfApproxPeoplePresent]];
    [mailBody appendFormat:@"\n"];
    
    NSString *intro1 = NSLocalizedString(@"EMAIL_INTRO_PART_ONE", @"From a total of \"$ 20,45\", which was spend on our last trip to \"Movies\". We all have to pay an equal share of \"$6,82\".");
    NSString *intro2 = NSLocalizedString(@"EMAIL_INTRO_PART_TWO", @"From a total of \"$ 20,45\", which was spend on our last trip to \"Movies\". We all have to pay an equal share of \"$6,82\".");
    NSString *intro3 = NSLocalizedString(@"EMAIL_INTRO_PART_THREE", @"From a total of \"$ 20,45\", which was spend on our last trip to \"Movies\". We all have to pay an equal share of \"$6,82\".");
    [mailBody appendFormat:@"%@ %@, %@ %@. %@ %@.\n", intro1, [nf stringFromNumber:[[self tonightsBill] totalSumOfMoneyOfThisSharedBill]], intro2,[[self tonightsBill] tripName], intro3, [nf stringFromNumber:[[self tonightsBill] amountPeopleShouldHavePaid]]];
    [mailBody appendFormat:@"\n"];
    
    if ([[self tonightsBill] totalAmountOfPeopleWhoHavePaid] == 0) {
        [mailBody appendFormat:@"%@\n", NSLocalizedString(@"EMAIL_NOBODY_HAS_PAID", @"The message that nobody has paid so far")];
    } else if ([[self tonightsBill] totalAmountOfPeopleWhoHavePaid] == 1) {
        [mailBody appendFormat:@"%@:\n", NSLocalizedString(@"EMAIL_ONE_PERSON_HAS_PAID", @"The person who has payed")];
    } else {
        [mailBody appendFormat:@"%@:\n", NSLocalizedString(@"EMAIL_MULTIPLE_PEOPLE_HAVE_PAID", @"The people who have paid are")];
    }
    NSSortDescriptor *sortDescriptor = [NSSortDescriptor sortDescriptorWithKey:@"dateCreated" ascending:YES];
    NSArray *allPayments = [[[self tonightsBill] payments] sortedArrayUsingDescriptors:@[sortDescriptor]];
    for (MCPayment *p in allPayments) {
        NSString *whoHasPaid1 = NSLocalizedString(@"EMAIL_WHO_HAS_PAID_ONE", @"Part one of the sentence: Mark has paid $24 for beer.");
        NSString *whoHasPaid2 = NSLocalizedString(@"EMAIL_WHO_HAS_PAID_TWO", @"Part two of the sentence: Mark has paid $24 for beer.");
        [mailBody appendFormat:@"%@ %@ %@ %@ %@.\n", [[p payingPerson] getName], whoHasPaid1, [nf stringFromNumber:[p money]], whoHasPaid2, [p descriptionOfPayment]];
    }
    [mailBody appendFormat:@"\n"];

//    NSString *averageOfUse = @"We agreed to pay an equal share on what we used. ";
    NSString *averageOfUse = NSLocalizedString(@"EMAIL_AVERAGE_OF_USE_SENTENCE_ONE", @"We agreed to pay an equal share on what we used.");
//    NSString *average1 = NSLocalizedString(@"EMAIL_AVERAGE_SENTENCES_ONE", @"Part one of: To have everybody pay the average of $7.00, I suggest the following solution:");
//    NSString *average2 = NSLocalizedString(@"EMAIL_AVERAGE_SENTENCES_TWO", @"Part two of: To have everybody pay the average of $7.00, I suggest the following solution:");
//    [mailBody appendFormat:@"%@ %@%@:\n", average1, [nf stringFromNumber:[[self tonightsBill] amountPeopleShouldHavePaid]], average2];
    [mailBody appendFormat:@"%@:\n", averageOfUse];

    for (MCPerson *person in [_tonightsBill peoplePresent]) {
        NSString *amountUsedString = [_tonightsBill amountShouldHavePaidAsCurrencyStringBy:person];
        NSString *usedString = NSLocalizedString(@"EMAIL_AMOUNT_USED", @"The word used in the sentence: Mark used $4.00");
        [mailBody appendFormat:@"%@ %@ %@.\n", [person getName], usedString, amountUsedString];
    }
    
    NSString *iSuggestString = NSLocalizedString(@"EMAIL_I_SUGGEST", @"I suggest the following solution.");
    [mailBody appendFormat:@"\n%@\n", iSuggestString];
    for (MCReturnPayment *rp in _solution) {
        [mailBody appendFormat:@"%@\n", [rp stringForMail]];
    }
    [mailBody appendFormat:@"\n"];
    [mailBody appendFormat:@"%@.", NSLocalizedString(@"EMAIL_FINAL SENTENCE", @"If you have any remarks please let me know.")];
    return mailBody;
}
@end

File: Going Dutch/Misc/MCMailComposer.swift
//
//  MCMailComposer.swift
//  We all pay
//
//  Created by Mark Cornelisse on 04/02/15.
//  Copyright (c) 2015 Mark Cornelisse. All rights reserved.
//

import Foundation

@objc(MCMailComposer)
class MCMailComposer: NSObject {
    var tonightsBill: MCSharedBill
    var isHTML: Bool = false
    
    @objc(initWithTonightsBill:)
    init(tonightsBill: MCSharedBill!) {
        self.tonightsBill = tonightsBill!
        super.init()
    }
    
    func getMailAddresses() -> [AnyObject]! {
        let allPeople = Array(tonightsBill.peoplePresent) as! [MCPerson]
        var listOfMailAddresses = [String]()
        for person in allPeople {
            if let emailAddress = person.defaultEmailAddress() {
                listOfMailAddresses.append(emailAddress)
            }
        }
        return listOfMailAddresses
    }
    
    func getSubject() -> String! {
        if let tripName = self.tonightsBill.tripName {
            return String.localizedStringWithFormat(NSLocalizedString("EMAIL_SUBJECT_WITH_TRIPNAME", comment: "Bill overview of our trip to %1$@"), tripName)
        } else {
            return NSLocalizedString("EMAIL_SUBJECT", comment: "Bill overview of our event.")
        }
    }
    
    func getMailBody() -> String! {
        let mainCurrencyFormatter = CurrencyFormatter()
        let localCurrencyFormatter = CurrencyFormatter()
        mainCurrencyFormatter.currencyCode = tonightsBill.mainCurrency.code
        
        let solution = tonightsBill.solveWhoHasToPayWhoFromThisBill() as! [ReturnPayment]
//        let sortDescriptorOnDateCreated = NSSortDescriptor(key: "dateCreated", ascending: true)
        let allPayments = Array(tonightsBill.payments) as! [MCPayment]
        let allPeople = Array(tonightsBill.peoplePresent) as! [MCPerson]
        
        var mailBody = String()
        mailBody += "https://itunes.apple.com/us/app/we-all-pay/id642135963?mt=8&uo=4\n\n"
        
        mailBody += String.localizedStringWithFormat(NSLocalizedString("EMAIL_DEAR", comment: "Dear %1$@,"), self.tonightsBill.stringOfApproxPeoplePresent())
        mailBody += "\n\n"
        
        if let tripName = self.tonightsBill.tripName {
            mailBody += String.localizedStringWithFormat(NSLocalizedString("EMAIL_INTRO_WITH_TRIPNAME", comment: "Here you go. The full overview of the %1$@ which we spend on our last event %2$@. We spent an average of %3$@ a person. You can find more of the details below."), mainCurrencyFormatter.stringForObjectValue(tonightsBill.totalSumOfMoneyOfThisSharedBill())!, tripName, mainCurrencyFormatter.stringForObjectValue(tonightsBill.amountPeopleShouldHavePaid())!)
        } else {
            mailBody += String.localizedStringWithFormat(NSLocalizedString("EMAIL_INTRO", comment: "Here you go. The full overview of the %1$@ which we spend on our last event. We spent an average of %2$@ a person. You can find more of the details below."), mainCurrencyFormatter.stringForObjectValue(tonightsBill.totalSumOfMoneyOfThisSharedBill())!, mainCurrencyFormatter.stringForObjectValue(tonightsBill.amountPeopleShouldHavePaid())!)
        }
        mailBody += "\n\n"
        
        switch tonightsBill.totalAmountOfPeopleWhoHavePaid() {
        case 1:
            mailBody += String.localizedStringWithFormat(NSLocalizedString("EMAIL_ONE_PERSON_HAS_PAID", comment: "The person who has payed %1$@"), allPayments.first!.payingPerson!.getFullName()!)
        case let totalAmountOfPeoplewhoHavePaid where totalAmountOfPeoplewhoHavePaid > 1:
            mailBody += String.localizedStringWithFormat(NSLocalizedString("EMAIL_MULTIPLE_PEOPLE_HAVE_PAID", comment: "The people who have paid are"))
        default:
            mailBody += String.localizedStringWithFormat(NSLocalizedString("EMAIL_NOBODY_HAS_PAID", comment: "The message that nobody has paid so far"))
        }
        mailBody += "\n"
        
        for payment in allPayments {
            if payment.exchangeRate.exchangeRate.doubleValue == 1.0 {
                mailBody += String.localizedStringWithFormat(NSLocalizedString("EMAIL_WHO_HAS_PAID", comment: "%1$@ has paid %2$@ for %3$@."), payment.payingPerson!.getFullName()!, mainCurrencyFormatter.stringForObjectValue(payment.moneyInMainCurrency())!, payment.fullDescriptionOfPayment()!)
            } else {
                localCurrencyFormatter.currencyCode = payment.currency.code
                mailBody += String.localizedStringWithFormat(NSLocalizedString("EMAIL_WHO_HAS_PAID_INTERNATIONAL", comment: "%1$@ has paid %2$@(%3$@) for %4$@."), payment.payingPerson!.getFullName()!, mainCurrencyFormatter.stringForObjectValue(payment.moneyInMainCurrency())!, localCurrencyFormatter.stringForObjectValue(payment.money)!, payment.fullDescriptionOfPayment()!)
            }
            mailBody += "\n"
        }
        mailBody += "\n"
        
        mailBody += String.localizedStringWithFormat(NSLocalizedString("EMAIL_AVERAGE_OF_USE_SENTENCE", comment: "The amount of money we used:"))
        mailBody += "\n"
        for person in allPeople {
            mailBody += String.localizedStringWithFormat(NSLocalizedString("EMAIL_AMOUNT_USED", comment: "%1$@ used %2$@ in total."), person.getFullName(), mainCurrencyFormatter.stringForObjectValue(tonightsBill.amountShouldHavePaidBy(person))!)
            mailBody += "\n"
        }
        mailBody += "\n"
        
        mailBody += String.localizedStringWithFormat(NSLocalizedString("EMAIL_I_SUGGEST", comment: "I suggest the following solution."))
        mailBody += "\n"
        
        for returnPayment in solution {
            let payerFullName: String = returnPayment.payer?.getFullName() ?? ""
            let moneyNumber = returnPayment.money ?? NSNumber(double: 0.0)
            let moneyString: String = mainCurrencyFormatter.stringForObjectValue(moneyNumber)!
            let receiverFullName: String = returnPayment.receiver?.getFullName() ?? ""
            mailBody += String.localizedStringWithFormat(NSLocalizedString("EMAIL_OWES", comment: "%1$@ pays %2$@ to %3$@"), payerFullName, moneyString, receiverFullName)
            mailBody += "\n"
        }
        mailBody += "\n"
        
        mailBody += NSLocalizedString("EMAIL_FINAL_SENTENCE", comment: "If you have any remarks please let me know.")
        mailBody += "\n"
        
        return mailBody
    }
}